File: ObjectiveC-DI/ObjectiveC-DI/APIClientImp.m
//
//  APIClientImp.m
//  ObjectiveC-DI
//
//  Created by Ilya Puchka on 28.02.16.
//  Copyright © 2016 Ilay Puchka. All rights reserved.
//

#import "APIClientImp.h"

@interface APIClientImp()

@property (nonatomic, copy) NSURL *baseURL;
@property (nonatomic, strong) id<NetworkSession> session;

@end

@implementation APIClientImp

- (instancetype)initWithBaseURL:(NSURL *)url session:(id<NetworkSession>)session {
    self = [super init];
    if (self != nil) {
        self.baseURL = url;
        self.session = session;
    }
    return self;
}

- (void)get:(NSString *)path {
    NSURLRequest *request = [self.session request:self.baseURL path:path];
    NSString *message = [NSString stringWithFormat:@"GET %@", request];
    [self.logger log:message];
}

@end

File: Swift-DI/Swift-DI/APIClientImp.swift
//
//  APIClientImp.swift
//  Swift-DI
//
//  Created by Ilya Puchka on 28.02.16.
//  Copyright © 2016 Ilay Puchka. All rights reserved.
//

import Foundation

class APIClientImp: NSObject, APIClient {
    
    let baseURL: NSURL
    let session: NetworkSession
    
    var logger: Logger?
    
    init(baseURL: NSURL, session: NetworkSession) {
        self.baseURL = baseURL
        self.session = session
    }

    func get(path: String) {
        let request = session.request(baseURL, path: path)
        let message = "GET \(request)"
        logger?.log(message)
    }
}
